patches/server/1029-Add-experience-points-API.patch

```diff
@@ -0,0 +1,58 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Lukas Planz <lukas.planz@web.de>
+Date: Tue, 5 Sep 2023 20:34:20 +0200
+Subject: [PATCH] Add experience points API
+
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+index 4e6fea7cf11b1e29ae7c7098a6f5d06bb5f93cc2..471b2015dfb7da62934d53fabbccf5e60f64ae06 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+@@ -1802,6 +1802,47 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+         Preconditions.checkArgument(exp >= 0, "Total experience points must not be negative (%s)", exp);
+         this.getHandle().totalExperience = exp;
+     }
++    // Paper start
++    @Override
++    public int getTotalExperiencePoints() {
++        return getTotalExperiencePoints(this.getLevel()) + Math.round(this.getExperiencePointsNeededForNextLevel() * getExp());
++    }
++
++    @Override
++    public void setTotalExperiencePoints(final int exp) {
++        int level = getLevelsForExperiencePoints(exp);
++        int remainingPoints = exp - getTotalExperiencePoints(level);
++
++        this.getHandle().experienceLevel = level;
++        this.getHandle().experienceProgress = (float) remainingPoints / this.getExperiencePointsNeededForNextLevel();
++        this.getHandle().lastSentExp = -1;
++    }
++
++    @Override
++    public int getExperiencePointsNeededForNextLevel() {
++        return this.getHandle().getXpNeededForNextLevel();
++    }
++
++    private int getTotalExperiencePoints(int level) {
```

Comment:
if we pull this, we def need a link to some documentation here
is there no way we could call a vanilla method that does something like this?

Reply:
Well, I pulled the equations from the [Minecraft Wiki](https://minecraft.fandom.com/wiki/Experience?so=search#Leveling_up) tbh. Sadly, the player experience is only stored as the level and the progression towards the next level as a float from 0.0 -> 1.0.

If you were to collect experience ingame the points are first converted into progression and then added to the current progression.

Reply:
just drop a link to that page in the code then, then people know where the magic numbers come from when reading the code